harness/determined/keras/callbacks.py

```diff
@@ -0,0 +1,642 @@
+            supports_tf_logs = getattr(cb, "_supports_tf_logs", False)
+            cb.on_predict_batch_end(batch, polylogs(supports_tf_logs))
+
+    def on_predict_end(self, logs: Optional[Dict] = None) -> None:
+        polylogs = _PolyLogs(logs)
+        for cb in self.callbacks:
+            supports_tf_logs = getattr(cb, "_supports_tf_logs", False)
+            cb.on_predict_end(polylogs(supports_tf_logs))
+
+    def _corrected_epoch_begin(self, epoch: int, logs: Dict) -> None:
+        """The real on_epoch_begin call, which is guaranteed to happen at the right time."""
+        polylogs = _PolyLogs(logs)
+        for cb in self.callbacks:
+            supports_tf_logs = getattr(cb, "_supports_tf_logs", False)
+            cb.on_epoch_begin(epoch, polylogs(supports_tf_logs))
+
+    def _corrected_epoch_end(self, epoch: int, logs: Dict) -> None:
+        """The real on_epoch_end call, which is guaranteed to happen at the right time."""
+        polylogs = _PolyLogs(logs)
+        for cb in self.callbacks:
+            supports_tf_logs = getattr(cb, "_supports_tf_logs", False)
+            cb.on_epoch_end(epoch, polylogs(supports_tf_logs))
+
+    def _corrected_test_end(self, logs: Dict) -> None:
+        polylogs = _PolyLogs(logs)
+        for cb in self.callbacks:
+            supports_tf_logs = getattr(cb, "_supports_tf_logs", False)
+            cb.on_test_end(polylogs(supports_tf_logs))
+
+    def _check_train_workload_begin(self) -> None:
+        if not self.want_train_workload_begin:
+            return
+        self.want_train_workload_begin = False
+        for cb in self.callbacks:
+            if isinstance(cb, Callback):
+                cb.on_train_workload_begin(self.state.total_batches, self.batches_requested, {})
+
+    def _check_epoch_begin(self) -> None:
+        if not self.state.want_epoch_begin:
+            return
+        self.state.want_epoch_begin = False
+        self._corrected_epoch_begin(self.state.epoch, {})
+
+    def _delayed_load_state(self) -> None:
+        """
+        Load state after on_train_begin, so that only the first on_train_begin can affect savable
+        state.  This is more consistent with how we don't allow on_train_end to alter savable
+        state, and also all TF-authored Callbacks would break if we didn't do it this way.
+        """
+        if self.load_state is None:
+            return
+        self.state = self.SavableState(**self.load_state["self"])
+        for cb, cb_state in zip(self.all_callbacks, self.load_state["callbacks"]):
+            if isinstance(cb, Callback):
+                cb.load_state(cb_state)
+
+    # Must be triggered externally.
+    def _get_state(self) -> Dict:
+        cb_state = [
+            cb.get_state() if isinstance(cb, Callback) else None for cb in self.all_callbacks
+        ]
+        return {"self": vars(self.state), "callbacks": cb_state}
+
+    # Must be triggered externally.
+    def _checkpoint_end(self, checkpoint_dir: str) -> None:
```

Comment:
non-blocking: can we make `checkpoint_dir` a `pathlib.Path` and convert it to a `str` when calling `on_checkpoint_end()`

Reply:
done